### PR TITLE
Fix amazingly slow boost iteration

### LINKED ISSUE #7639: EnumerateLibrary.GetPosition() surprisingly slow
**Describe the bug**
In a simple enumeration loop, if one is iterating over library.GetPosition() in the loop, it takes up to 40% of the runtime of the loop.
For a full enumeration, counting/calculating the position yourself everytime in python is much faster and reduces the overhead to almost zero compared to the actual work in the iterator (i.e. RunReactants).

**To Reproduce**
Will update to a full example soon.

```
    library = EnumerateLibrary(rxn, bbs)
    reagents = library.GetReagents()
    for product_template_results in library:
        # print input reactant monomers and the product
        libpos = library.GetPosition()
        for i, n in enumerate(libpos): # <- this line iterates over a vector of size_t with 3 elements and takes 40% of the runtime
            print(reagents[i][n].GetProp("_Name"))
```

**Expected behavior**
The line taking less than 1% of the runtime.

**Screenshots**
Screenshot 2024-07-20 at 11 26 53

**Configuration (please complete the following information):**
 - RDKit version: 2023.09.6
 - OS: macOS 14.5
 - Python version (if relevant): 3.11
 - Are you using conda? no

### PR DESCRIPTION
**Pull request by bp-kelley**
Fixes #7639 

This is a simple fix to return a list as opposed to a wrapped boost vector for GetPosition.  Iteration through the latter is remarkably slow.

### CODE REVIEW

**Comment by bp-kelley** [comment]
@greglandrum I might check to see if we return other  directly referenced vectors, indexing them is fine, but using them in for loops is stupid slow.

**Comment by greglandrum** [comment]
> @greglandrum I might check to see if we return other directly referenced vectors, indexing them is fine, but using them in for loops is stupid slow.

Indexing them is actually also not fine. I added some notes to the original issue.

**Comment by greglandrum** [comment]
Rather than applying this one bandaid, I think it would make sense to put a more general `vectorToTuple()` function in place and use that to convert the return values. We'll still need to write individual helpers like you did here, but at least we can factor out the loop over the 

I also think the return type should be a tuple and not a list. The objects returned are immutable.

**Comment by bp-kelley** [comment]
I'll look at this.

**Comment by bp-kelley** [comment]
I was wrong, it's super simple, we can return tuples and lists from the initializers.

```
return python::tuple(base->getPosition());
return python::list(base->getPosition());
```

So we don't need the vectorToTuple or vectorToList.

**Comment by bp-kelley** [comment]
> > @greglandrum I might check to see if we return other directly referenced vectors, indexing them is fine, but using them in for loops is stupid slow.
> 
> Indexing them is actually also not fine. I added some notes to the original issue.

I meant "fine" in the sense it was the iteration that was slow, not that it was acceptable.

**Comment by bp-kelley** [comment]
I threw in a sequenceToList implementation.  The tuple version I haven't figured out yet, but here are the result for the path length test:

```
before
318 μs ± 1.82 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

after
1.45 μs ± 9.15 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

**Comment by bp-kelley** [comment]
I think for complicated results, std::vector<std::vector<T>> python lists will be way easier and are faster to create anyway by 1.5 to 3x from the python CAPI it looks like.

**Comment by greglandrum** [comment]
> I think for complicated results, std::vector<std::vector> python lists will be way easier and are faster to create anyway by 1.5 to 3x from the python CAPI it looks like.

Why python lists instead of tuples? The results are immutable, so it seems like tuples make more sense. Is there a massive performance penalty associated with converting the python::list to a python::tuple on the C++ side?

File: Code/RDBoost/Wrap.h [hunk 1]
```diff
@@ -370,4 +370,53 @@ struct rdkit_pickle_suite : python::pickle_suite {
...
 };
 
+template<typename T>
+void addToList(python::list &l, const T &v) {
+  l.append(v);
+}
+
+template<typename T>
+void addToList(python::list &l, const std::vector<T> &v) {
+  l.append(python::list(v));
```
Comment (bp-kelley) [inline Code/RDBoost/Wrap.h:380]: This should probably be 
```
addToList<T>(l, v)
```

And then we could get rid of the std::vector<std::vector<T>> funds

File: Code/RDBoost/Wrap.h [hunk 2]
```diff
@@ -370,4 +370,53 @@ struct rdkit_pickle_suite : python::pickle_suite {
   static bool getstate_manages_dict() { return true; }
 };
 
+template<typename T>
```
Comment (bp-kelley) [inline Code/RDBoost/Wrap.h:373]: I don't think this can ever be called.

**Comment by bp-kelley** [comment]
I did the test, no noticeable difference between list and tuple

1.45 μs ± 7.66 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

**Comment by bp-kelley** [comment]
If you're okay with this, I can start changing the other functions that return boost wrapped vectors directly

**Comment by greglandrum** [comment]
> If you're okay with this, I can start changing the other functions that return boost wrapped vectors directly

I only think we want to modify functions that are returning an object, not a reference (assuming we have any of those). The reference probably should remain mutable.

**Comment by greglandrum** [comment]
I did some more exploring this morning.

We can mostly (comments below on when this doesn't work) get away with this:
```
// From boost::python examples
template <typename T>
struct iterable_to_tuple {
  static PyObject *convert(T const &p) {
    python::list l;
    for (auto const &elem : p) {
      l.append(elem);
    }
    return boost::python::incref(boost::python::tuple(l).ptr());
  }

  static PyTypeObject const *get_pytype() { return &PyTuple_Type; }
};

// Helper for convenience.
template <typename T>
struct iterable_to_python_converter {
  iterable_to_python_converter() {
    if (is_python_converter_registered<T>()) {
      std::cerr
          << "Warning: iterable_to_python_converter: Python converter already registered for type "
          << typeid(T).name() << std::endl;
      return;
    }

    boost::python::to_python_converter<T, iterable_to_tuple<T>,
                                       true  // we have get_pytype
                                       >();
  }
};
```
And then replacing `RegisterVectorConverter<int>()` with `iterable_to_python_converter<std::vector<in>>()`

The results are, as you've observed, slower to construct (because all the `to_python()` stuff happens initially), but much faster to work with.

I benchmarked it by exposing these simple functions:
```
std::vector<int> test_vector(unsigned int sz) {
  std::vector<int> v(sz, 0);
  return v;
}
std::vector<std::vector<int>> test_vector_vector(unsigned int sz1,
                                                 unsigned int sz2) {
  std::vector<std::vector<int>> v(sz1, std::vector<int>(sz2, 0));
  return v;
}
const std::vector<unsigned int> test_uvector(unsigned int sz) {
  std::vector<unsigned int> v(sz, 0);
  return v;
}
```
and then testing with:
```
from rdkit import rdBase
from IPython import get_ipython

ipy = get_ipython()
print('return vector(10)')
ipy.run_line_magic('timeit', 'obj = rdBase.test_vector(10)')
print('return vector_vector(10,10)')
ipy.run_line_magic('timeit', 'rdBase.test_vector_vector(10,10)')
print('iterate vector(10)')
obj = rdBase.test_vector(10)
ipy.run_line_magic('timeit', '[x for x in obj]')
print('iterate vector_vector(10,10)')
obj = rdBase.test_vector_vector(10, 10)
ipy.run_line_magic('timeit', '[[y for y in x] for x in obj]')
print('index vector(10)')
obj = rdBase.test_vector(10)
ipy.run_line_magic('timeit', '[obj[i] for i in range(len(obj))]')
print('index vector_vector(10,10)')
obj = rdBase.test_vector_vector(10, 10)
ipy.run_line_magic('timeit', '[[obj[i][j] for j in range(len(obj[i]))] for i in range(len(obj))]')
```
Using the current wrapping method with the `VectorIndexingSuite` I get:
```
return vector(10)
74.9 ns ± 0.246 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
return vector_vector(10,10)
393 ns ± 0.768 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
iterate vector(10)
4.76 µs ± 7.48 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
iterate vector_vector(10,10)
53.6 µs ± 100 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
index vector(10)
913 ns ± 11.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
index vector_vector(10,10)
39.5 µs ± 206 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
Returning tuples:
```
return vector(10)
158 ns ± 1.03 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
return vector_vector(10,10)
1.53 µs ± 7.46 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
iterate vector(10)
117 ns ± 0.827 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
iterate vector_vector(10,10)
1.38 µs ± 7.65 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
index vector(10)
203 ns ± 0.208 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
index vector_vector(10,10)
2.85 µs ± 9.77 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

For completeness, I also modified the code to return lists, this is a bit faster to return the results, but has no other performance impact:
```
return vector(10)
110 ns ± 0.803 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
return vector_vector(10,10)
1.13 µs ± 8.86 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
iterate vector(10)
118 ns ± 0.754 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
iterate vector_vector(10,10)
1.39 µs ± 9.22 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
index vector(10)
195 ns ± 5.88 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
index vector_vector(10,10)
2.93 µs ± 11.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

I think the extra time to return a tuple will not be noticeable if we're working with a function that's actually doing something and that it's worth it in order to make it clear that the result object is immutable (i.e. modifying it has no impact on the underlying C++ objects)

Using this approach will require some code changes.
For example, I have not been able to figure out how to make it work with object attributes that are exposed using `.def_readonly()`, like here:
https://github.com/rdkit/rdkit/blob/e5320328d74f383027fb2693f90250c1c0aff012/Code/GraphMol/Wrap/Chirality.cpp#L58

For that I had to write a simple helper function:
```
std::vector<unsigned> getControllingAtoms(const Chirality::StereoInfo &self) {
  return self.controllingAtoms;
}
```
and then expose that:
```
        .add_property("controllingAtoms", getControllingAtoms,
                      "indices of the atoms controlling the stereo");
```

There are only a few places where we do this, so it's pretty minimal work. I think it's worth trying out.

@bp-kelley what do you think?

**Comment by bp-kelley** [comment]
The the necessary helper functions, I think just returning `python::tuple(self.controllingAtoms)` might be better.

I expect this doesn't handle std::vector<std::vector<int>> which is used a lot, but we could make another converter for that I suppose.  Do you want to add these to this branch so we can play around with them before the final decision is made?  I think registering converters is the way to go, btw.

**Comment by bp-kelley** [comment]
And python::list and python::tuple takes a sequence so:

```
static PyObject *convert(T const &p) {
    return boost::python::incref(boost::python::tuple(p).ptr());
  }
```

should work as well.

**Comment by greglandrum** [comment]
> And python::list and python::tuple takes a sequence so:
> 
> ```
> static PyObject *convert(T const &p) {
>     return boost::python::incref(boost::python::tuple(p).ptr());
>   }
> ```
> 
> should work as well.

I'm pretty sure I tried that and it did not work

**Comment by bp-kelley** [comment]
Go for it

**Comment by bp-kelley** [comment]
I'll have time to take another look at this later this week, hopefully something concrete by next.

**Comment by greglandrum** [comment]
@bp-kelley are you going to have time to come back to this?

**Comment by bp-kelley** [comment]
In testing phase now

**Comment by bp-kelley** [comment]
@greglandrum cross your fingers.  I had to implement custom from/to converters for the vectors.  I can make the to python's a bit better now, but I ran out of steam :)

The one big change I made to your code was enabling better conversion to std::vector<std::vector<int>> which created a second converter class.

**Comment by bp-kelley** [comment]
The other issue that took me a bit is the custom python functions that take const std::vector<unsigned int> & which is currently not supported in the pyfunctor for some reason I don't understand yet.

**Comment by greglandrum** [comment]
I think this last set of changes goes too far: there are times when we need the vector_indexing_suite because it provides an mutable interface to the underlying C++ object.

Here's one example of that (it's a sillly one, but there are likely other parts of the code where this behavior is important).

from the current release branch (it's the same on master):
```
In [27]: m = Chem.MolFromSmiles('C[C@@H]1CC(C[C@@H](C)[C@@H](C)O)C[C@H](C)C1 |o1:1,7|')

In [28]: si = Chem.FindPotentialStereo(m)

In [29]: si[0].controllingAtoms
Out[29]: <rdkit.rdBase._vectj at 0x7f763bef7760>

In [30]: list(si[0].controllingAtoms)
Out[30]: [0, 2, 13]

In [31]: si[0].controllingAtoms.append(6)

In [32]: list(si[0].controllingAtoms)
Out[32]: [0, 2, 13, 6]
```

and on your branch:
```
In [6]: m = Chem.MolFromSmiles('C[C@@H]1CC(C[C@@H](C)[C@@H](C)O)C[C@H](C)C1 |o1:1,7|')

In [7]: si = Chem.FindPotentialStereo(m)

In [9]: si[0]
Out[9]: <rdkit.Chem.rdchem.StereoInfo at 0x7a31955f3740>

In [10]: si[0].controllingAtoms
Out[10]: [0, 2, 13]

In [11]: si[0].controllingAtoms.append(3)

In [12]: si[0].controllingAtoms
Out[12]: [0, 2, 13]
```

It's safe to change how we return sequences which are either not references, or which are const references, but the cases where we give Python a non-const reference to a `std::vector` should continue to use the existing converters.

**Comment by bp-kelley** [comment]
It is unclear to me if you can have both the vector indexing suite only for std::vector<T>& but I can give it a try.

**Comment by bp-kelley** [comment]
Replacing RegisterVectorConverter with the to_python_converter is what disables this btw.

I'll see if we can get it to work with references.

**Comment by bp-kelley** [comment]
After looking at the code, I'm a bit confused.  controllingAtoms was set to a "readonly" prop

```
.def_readonly("controllingAtoms",
                       &Chirality::StereoInfo::controllingAtoms,
```

I wouldn't expect it to be modifiable.  Do we really have any .def_readwrites, is that our intention?

**Comment by bp-kelley** [comment]
Ok, from an old post from Dave Abrahams
```
Well, maybe the doc should be clearer, on this topic, but def_readonly
makes the *attribute* read-only or read-write. Python has no
distinction between const and non-const; it only has mutable and
immutable objects. The map you've exposed is mutable.
```

**Comment by greglandrum** [comment]
> Ok, from an old post from Dave Abrahams
> 
> ```
> Well, maybe the doc should be clearer, on this topic, but def_readonly
> makes the *attribute* read-only or read-write. Python has no
> distinction between const and non-const; it only has mutable and
> immutable objects. The map you've exposed is mutable.
> ```

Exactly. You can't say `si.controllingAtoms = foo`, but you can modify the `controllingAtoms` once you have them

I just provided that specific example because it was one that I had noticed earlier. The general point is the need to return non-const `std::vector` references to python as mutable objects

**Comment by bp-kelley** [comment]
I'm not sure we can have our cake and eat it too.   If we make a new python converter we have to use it for everything which probably means we go back to the initial solution of using helpers that return tuples.

**Comment by greglandrum** [comment]
> I'm not sure we can have our cake and eat it too. If we make a new python converter we have to use it for everything which probably means we go back to the initial solution of using helpers that return tuples.

I think we ought to be able to have the automatic conversion to tuple (not to list) and then add helper functions that return something built using the vector_indexing_suite for the cases where we need to return something mutable.

**Comment by bp-kelley** [comment]
It's the same return type which might make it tricky.

**Comment by bp-kelley** [comment]
Maybe you can return a vector_indexing_suite wrapped instead...

**Comment by greglandrum** [comment]
> It's the same return type which might make it tricky.

yeah, I think it's going to require specific helper functions for each case where we need to do this (there aren't that many of them)

**Comment by bp-kelley** [comment]
Can't say I'm proud of myself, but I have a solution...

I made a thin proxy around std::vectors so we can run the indexing suite on that as opposed to the underlying vector.

File: Code/RDBoost/Wrap.h (hunk 2 above)
Comment (greglandrum) [reply Code/RDBoost/Wrap.h:379]: I'm not sure what you mean, it's used below by the vec_to_python converter.

**Comment by greglandrum** [comment]
@bp-kelley https://github.com/bp-kelley/rdkit/pull/25 has my (hopefully) last set of suggested changes.

**Comment by bp-kelley** [comment]
I beginning to think we should just try and fix the enumerator if the vector indexing suite.

Test 1, retrieve and convert library.GetPosition() in a list.

Original results:

```
retrieve and convert to list: list(library.GetPosition())
10000 loops: best of 5: 409 usec per loop

raw access: [p[i] for i in range(3)]
10000 loops: best of 5: 543 nsec per loop

retrieve and raw access: p=library.GetPosition();[p[i] for i in range(3)]
10000 loops: best of 5: 1.1 usec per loop
```

new results:

```
retrieve and convert to list: list(library.GetPosition())
10000 loops: best of 5: 187 usec per loop

raw access: [p[i] for i in range(3)]
10000 loops: best of 5: 176 nsec per loop

retrieve and raw access: p=library.GetPosition();[p[i] for i in range(3)]
10000 loops: best of 5: 186 usec per loop
```

So the fastest current way for "small vectors" like GetPositions is to use the vector indexing suite but not use the enumerator.  Maddening!

**Comment by greglandrum** [comment]
I created a benchmarking example where `GetPosition()` has a length of 20:
```
from rdkit import Chem
from rdkit.Chem import rdChemReactions

r_dumb = []
p_dumb = []
reagents = []
for i in range(1, 11):
  r_dumb.append(f'[N:{i*10}3]=[C:{i*10}1].[N:{i*10}2]')
  p_dumb.append(f'[N:{i*10}3]-[C:{i*10}1]-[N+0:{i*10}2]')
  reagents.append([Chem.MolFromSmiles('C=CCN=C=S'), Chem.MolFromSmiles('CC=CCN=C=S')])
  reagents.append([
    Chem.MolFromSmiles('NCc1ncc(Cl)cc1Br'),
    Chem.MolFromSmiles('NCCc1ncc(Cl)cc1Br'),
    Chem.MolFromSmiles('NCCCc1ncc(Cl)cc1Br')
  ])
r_dumb = '.'.join(r_dumb)
p_dumb = '.'.join(p_dumb)

smirks_dumb = "[N:3]=[C:1].[N:2]>>[N:3]-[C:1]-[N+0:2]"
rxn = rdChemReactions.ReactionFromSmarts(f'{r_dumb}>>{p_dumb}')

enumerator = rdChemReactions.EnumerateLibrary(rxn, reagents)

from IPython import get_ipython

ipy = get_ipython()

ipy.run_line_magic('timeit', 'ps = enumerator.GetPosition()')
ipy.run_line_magic('timeit', 'ps = enumerator.GetPosition();i = iter(ps)')
ipy.run_line_magic('timeit', 'ps = enumerator.GetPosition();o = next(iter(ps))')
ipy.run_line_magic('timeit', 'list(enumerator.GetPosition())')
ipy.run_line_magic('timeit', 'ps = [x for i,x in enumerate(enumerator.GetPosition())]')

ipy.run_line_magic('timeit', 'ps = enumerator.GetPosition();[ps[i] for i in range(len(ps))]')

ipy.run_line_magic('timeit',
                   'ps = enumerator.GetPosition();lps=len(ps);[ps[i] for i in range(lps)]')
```
Running this with master produces:
```
421 ns ± 0.987 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
807 ns ± 8.23 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.09 µs ± 5.72 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
16.2 µs ± 40.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.6 µs ± 48.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
5.7 µs ± 17.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
5.73 µs ± 15.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
Running it with this branch produces:
```
16.5 µs ± 176 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.6 µs ± 169 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.5 µs ± 9.57 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.6 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17 µs ± 40.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17 µs ± 17.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17.3 µs ± 227 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
So the current status is that this branch is more or less the same speed for `ps = [x for i,x in enumerate(enumerator.GetPosition())]` and slower (sometimes significantly slower) for all the other use cases.

**Comment by bp-kelley** [comment]
Yep, let’s close this one as an interesting experiment…

**Comment by bp-kelley** [comment]
#replaced by #7719

**Pull request closed without merging**